### PR TITLE
Bugfix/qrcode version estimator

### PR DESCRIPTION
**Pull request by bagelbytes61**
Fixes https://github.com/opencv/opencv/issues/24366

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [x] There is a reference to the original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by dkurt** [comment]
Please rebase source branch

**Comment by bagelbytes61** [comment]
> Please rebase source branch

Done!

**Comment by AleksandrPanov** [comment]
@bagelbytes61, one test crashes:

```
Objdetect_QRCode_Encode_Decode.regression :
unknown file: Failure
C++ exception with description "OpenCV(4.8.0-dev) /build/precommit_linux64/4.x/opencv/modules/objdetect/src/qrcode_encoder.cpp:371: error: (-5:Bad argument) The given version is not suitable for the given input string length  in function 'generateQR'
" thrown in the test body.
```

https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/104823

https://pullrequest.opencv.org/buildbot/builders/precommit_linux64/builds/104823/steps/test_objdetect/logs/tests%20summary

There were also warnings:
```
/build/precommit_linux64/4.x/opencv/modules/objdetect/src/qrcode_encoder.cpp:681:12: warning: enumeration value 'MODE_AUTO' not handled in switch [-Wswitch]
/build/precommit_linux64/4.x/opencv/modules/objdetect/src/qrcode_encoder.cpp:681:12: warning: enumeration value 'MODE_ECI' not handled in switch [-Wswitch]
/build/precommit_linux64/4.x/opencv/modules/objdetect/src/qrcode_encoder.cpp:681:12: warning: enumeration value 'MODE_KANJI' not handled in switch [-Wswitch]
/build/precommit_linux64/4.x/opencv/modules/objdetect/src/qrcode_encoder.cpp:681:12: warning: enumeration value 'MODE_STRUCTURED_APPEND' not handled in switch [-Wswitch]
```

**Comment by AleksandrPanov** [comment]
```
[  FAILED  ] 4 tests, listed below:
[  FAILED  ] Objdetect_QRCode_Encode_Decode.regression
[  FAILED  ] Objdetect_QRCode_Encode.regression/3, where GetParam() = "version2_mode1.png"
[  FAILED  ] Objdetect_QRCode_Encode.regression/4, where GetParam() = "version2_mode2.png"
[  FAILED  ] Objdetect_QRCode_Encode.regression/6, where GetParam() = "version3_mode2.png"
```

**Comment by asmorkalov** [comment]
@bagelbytes61 Thanks for the patch. The fix changes version for several test cases, where version is set to auto (0). You need to create PR with the same branch name as this one to OpenCV extra and update reference images. Example:
```
[ RUN      ] Objdetect_QRCode_Encode.regression/3, where GetParam() = "version2_mode1.png"
[DEBUG:0@0.152] global qrcode_encoder.cpp:370 generateQR QR code detected_version: 4 version_level: 0
src resolution: 29x29
result resolution: 37x37
unknown file: Failure
C++ exception with description "OpenCV(4.8.0-dev) /home/alexander/Projects/OpenCV/opencv-master/modules/core/src/arithm.cpp:652: error: (-209:Sizes of input arguments do not match) The operation is neither 'array op array' (where arrays have the same size and the same number of channels), nor 'array op scalar', nor 'scalar op array' in function 'arithm_op'
" thrown in the test body.
[  FAILED  ] Objdetect_QRCode_Encode.regression/3, where GetParam() = "version2_mode1.png" (76 ms)
```

**Comment by asmorkalov** [comment]
@dkurt Could you take a look too? Loos like the PR overestimate minimal version of QR code for several test cases. The QR codes in our test data contains full (not truncated) sequence.

**Comment by bagelbytes61** [comment]
> @bagelbytes61 Thanks for the patch. The fix changes version for several test cases, where version is set to auto (0). You need to create PR with the same branch name as this one to OpenCV extra and update reference images. Example:
> 
> ```
> [ RUN      ] Objdetect_QRCode_Encode.regression/3, where GetParam() = "version2_mode1.png"
> [DEBUG:0@0.152] global qrcode_encoder.cpp:370 generateQR QR code detected_version: 4 version_level: 0
> src resolution: 29x29
> result resolution: 37x37
> unknown file: Failure
> C++ exception with description "OpenCV(4.8.0-dev) /home/alexander/Projects/OpenCV/opencv-master/modules/core/src/arithm.cpp:652: error: (-209:Sizes of input arguments do not match) The operation is neither 'array op array' (where arrays have the same size and the same number of channels), nor 'array op scalar', nor 'scalar op array' in function 'arithm_op'
> " thrown in the test body.
> [  FAILED  ] Objdetect_QRCode_Encode.regression/3, where GetParam() = "version2_mode1.png" (76 ms)
> ```

Apologize for not getting back to you sooner. I addressed the test case failures except for the ECI  case is still failing. Would you be able to provide any insight into the failure? I don't know much about QR code technology, so if there's a technical reason for the failures I am afraid that it may be a bit over my head...

File: modules/objdetect/src/qrcode_encoder.cpp [hunk 1]
```diff
@@ -268,53 +269,67 @@ int QRCodeEncoderImpl::findVersionCapacity(const int input_length, const int ecc
...
+        1, // cv::QRCodeEncoder::EncodeMode::MODE_ALPHANUMERIC
+       -1,
+        2, // cv::QRCodeEncoder::EncodeMode::MODE_BYTE
+       -1,
+       -1,
+       -1,
+        3  // cv::QRCodeEncoder::EncodeMode::MODE_KANJI
+    };
+
+    if (input_length > version_capacity_database[40].ec_level[ecc_level].encoding_modes[mode_index[mode]])
```
Comment (dkurt) [inline modules/objdetect/src/qrcode_encoder.cpp:290]: `version_capacity_database[MAX_VERSION]`

File: modules/objdetect/src/qrcode_encoder.cpp [hunk 2]
```diff
@@ -268,53 +269,67 @@ int QRCodeEncoderImpl::findVersionCapacity(const int input_length, const int ecc
...
     {
-        possible_version.push_back(1);
-        possible_version.push_back(2);
+        return false;
     }
-    else if (input_length <= version_capacity_database[26].ec_level[ecc_level].encoding_modes[3])
+
+    int version = 40;
```
Comment (dkurt) [inline modules/objdetect/src/qrcode_encoder.cpp:295]: Use `MAX_VERSION` from https://github.com/opencv/opencv/blob/c2f909fc86cae0e44ccd9a27dd52578682920f29/modules/objdetect/src/qrcode_encoder_table.inl.hpp#L14

File: modules/objdetect/src/qrcode_encoder.cpp [hunk 3]
```diff
@@ -633,14 +651,61 @@ bool QRCodeEncoderImpl::isAlphaNumeric(const std::string& input)
...
+        default:
+            break;
+    }
+
+    if (mode != nullptr)
+    {
+        *mode = selected_mode;
+    }
+
     return true;
```
Comment (dkurt) [inline modules/objdetect/src/qrcode_encoder.cpp:709]: There is no option for this method to return `false`

File: modules/objdetect/src/qrcode_encoder.cpp [hunk 4]
```diff
@@ -268,53 +269,67 @@ int QRCodeEncoderImpl::findVersionCapacity(const int input_length, const int ecc
...
-    else
+
+    possible_version.push_back(version);
+
+    if (version < MAX_VERSION)
     {
-        possible_version.push_back(3);
+        possible_version.push_back(version + 1);
     }
+
```
Comment (dkurt) [inline modules/objdetect/src/qrcode_encoder.cpp:315]: Two similar if statements `if (version < MAX_VERSION)`

File: modules/objdetect/src/qrcode_encoder.cpp [hunk 5]
```diff
@@ -342,19 +360,22 @@ void QRCodeEncoderImpl::generateQR(const std::string &input)
...
 
         payload.clear();
         payload.reserve(MAX_PAYLOAD_LEN);
         format = vector<uint8_t> (15, 255);
         version_reserved = vector<uint8_t> (18, 255);
-        version_size = (21 + (version_level - 1) * 4);
-        version_info = makePtr<VersionInfo>(version_info_database[version_level]);
+        version_size = (21 + (tmp_version_level - 1) * 4);
+        version_info = makePtr<VersionInfo>(version_info_database[tmp_version_level]);
         cur_ecc_params = makePtr<BlockParams>(version_info->ecc[ecc_level]);
```
Comment (asmorkalov) [inline modules/objdetect/src/qrcode_encoder.cpp:379]: No need to use Ptr here.

File: modules/objdetect/src/qrcode_encoder.cpp (hunk 4 above)
Comment (asmorkalov) [reply modules/objdetect/src/qrcode_encoder.cpp:319]: Why do not push all versions between `version` and `MAX_VERSION`?

File: modules/objdetect/src/qrcode_encoder.cpp [hunk 6]
```diff
@@ -247,13 +248,13 @@ int QRCodeEncoderImpl::eccLevelToCode(CorrectionLevel level)
...
+int QRCodeEncoderImpl::findVersionCapacity(const int input_length, const int ecc, const std::vector<int>& possible_versions)
 {
     int data_codewords, version_index = -1;
     const int byte_len = 8;
     version_index = -1;
 
-    for (int i = version_begin; i < version_end; i++)
+    for (int i : possible_versions)
     {
         Ptr<BlockParams> tmp_ecc_params = makePtr<BlockParams>(version_info_database[i].ecc[ecc]);
```
[the remark this answers is not in the document]
Comment (dkurt) [reply modules/objdetect/src/qrcode_encoder.cpp]: Sure, replaced

**Review by asmorkalov — APPROVED** [review]
:+1:

**Review by AleksandrPanov — COMMENTED** [review]
👍

**Pull request merged**